src/structures/paging/recursive.rs

```diff
@@ -237,6 +276,224 @@ impl<'a> RecursivePageTable<'a> {
+        &mut self,
+        range: R,
+        deallocator: &mut D,
+        mode: ReclaimPageTablesMode,
+    ) where
+        D: FrameDeallocator<Size4KiB>,
+        R: RangeBounds<Page<Size4KiB>>,
+    {
+        // Get the page ranges.
+        let start = match range.start_bound() {
+            Bound::Included(endpoint) => *endpoint,
+            Bound::Excluded(endpoint) => *endpoint + 1,
+            Bound::Unbounded => {
+                panic!("Attempt to reclaim page tables for unbounded region of memory.")
+            }
+        };
+        let end = match range.end_bound() {
+            Bound::Included(endpoint) => *endpoint + 1,
+            Bound::Excluded(endpoint) => *endpoint,
+            Bound::Unbounded => {
+                panic!("Attempt to reclaim page tables for unbounded region of memory.")
+            }
+        };
+
+        let p4 = &mut self.p4;
+
+        // Starting at the leaf page tables (p1), we reclaim page tables. Then we do p2 and p3.
+
+        // We can only reclaim a page table if all of its possible children are in the given range,
+        // so round `start` and `end` to exclude page tables that are only partial in range.
+        //
+        // That is, we round `start` up to the nearest 512 * Size4KiB boundary, and we round `end`
+        // down to the nearest boundary.
+
+        let p1_start = if start.p1_index() == u9::new(0) {
+            start
+        } else {
+            Page::from_page_table_indices(
+                start.p4_index(),
+                start.p3_index(),
+                start.p2_index(),
+                start.p1_index() + u9::new(1),
+            )
+        };
+
+        let p1_end = Page::from_page_table_indices(
+            start.p4_index(),
+            start.p3_index(),
+            start.p2_index(),
+            u9::new(0),
+        );
+
+        // Free all the page tables!
+        for page in p1_start..p1_end {
+            let p4_entry = &mut p4[page.p4_index()];
+
+            match p4_entry.frame() {
+                Err(FrameError::FrameNotPresent) => {
+                    continue;
+                }
+                Err(FrameError::HugeFrame) => unreachable!(),
+                _ => {}
+            };
+
+            let p3 = &mut *(p3_ptr(page, self.recursive_index));
+            let p3_entry = &mut p3[page.p3_index()];
+
+            match p3_entry.frame() {
+                Err(FrameError::FrameNotPresent) => {
+                    continue;
+                }
+                Err(FrameError::HugeFrame) => unreachable!(),
+                _ => {}
+            };
+
+            let p2 = &mut *(p2_ptr(page, self.recursive_index));
+            let p2_entry = &mut p2[page.p2_index()];
+
+            let p1_frame = match p2_entry.frame() {
+                Ok(frame) => frame,
+                Err(FrameError::FrameNotPresent) => {
+                    continue;
+                }
+                Err(FrameError::HugeFrame) => unreachable!(),
```

Comment:
Why is this unreachable? 2MiB pages are totally possible.

Reply:
...or here...